enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,877 @@
+    For setting dynamic header values for http request. Allowed sample fetchers are req.hdr, ssl_c_der. Allowed converters are lower, base64.
+    Example: One might be able to set value of `x-forwarded-client-cert` to `%{+Q}[ssl_c_der,base64]`.
+
+- Certain headers like HSTS are not allowed to be set using this API as there is already a route annotation available for this
+  Would it not be better to provide a concrete API and deprecate the annotation API?
+  Answer: The strict-transport-security header (AKA HSTS) has wider ranging implications than the typical HTTP header. We considered allowing HSTS to be configured
+    on the IngressController in the past; see the global-options-enable-hsts EP, which was never implemented. However, HSTS isn't neatly scoped: Route A's HSTS
+    policy can affect Route B, even if the routes are in different namespaces or belong to different IngressControllers, if Route B's host name is a subdomain of
+    Route A's. Additionally, specifying HSTS on an IngressController can cause a route manifest that behaves as expected on one cluster to behave surprisingly and
+    problematically on another cluster. For these reasons, we decided not to implement the global-options-enable-hsts EP and instead wrote and implemented the
+    global-admission-hsts EP, which defines a cluster-scoped API, using which the cluster admin can specify HSTS parameters that may or must be specified on routes
+    for a given domain: https://github.com/openshift/api/blob/6ba31fa438f20d6c822ef47ad3d771309cb5216b/config/v1/types_ingress.go#L64-L87. Based on this experience,
+    we wouldn't want to allow configuring HSTS on the IngressController using this new API.
+
+### Test Plan
+
+The controller that manages the IngressController Deployment and related
+resources has unit test coverage; for this enhancement, the unit tests are
+expanded to cover the additional functionality.
+
+The operator has end-to-end tests; for this enhancement, a test is added that
+(1) creates an IngressController that specifies certain response
+headers should be set or deleted, (2) sends an HTTP request to one of the standard
+routes (such as the console route), and (3) verifies that the expected headers
+are seen in the `curl -vk <route host>`.
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+#### Tech Preview -> GA
+
+#### Removing a deprecated feature
+
+### Upgrade / Downgrade Strategy
+
+On upgrade, the default configuration sets or deletes no HTTP headers, as before the
+introduction of the HTTP header Set/Delete API.
+
+On downgrade, the HAProxy configuration template ignores unrecognized
+environment variables set by ingress operator using the value set in the Ingress Controller CR and
+the ServiceAlias config struct variable set by router using the value set in the routeSpec.
+
+### Version Skew Strategy
+
+### Operational Aspects of API Extensions
+
+#### Failure Modes
+
+#### Support Procedures
+
```

Comment:
These sections are still empty.  Hopefully we can avoid most of the possible failure modes through validation.  I suppose one failure mode would be exceeding `tune.bufsize`, or setting a header that confuses some server or some Web browser.  Support procedures could include checking access logs for client or server errors and checking for HAProxy config parse errors.

Reply:
somewhat github hid this comment not sure why. anyway will add these.

Reply:
tune.maxrewrite {{ env "ROUTER_MAX_REWRITE_SIZE" "8192" }}
  tune.bufsize {{ env "ROUTER_BUF_SIZE" "32768" }

If an HTTP request is larger than (tune.bufsize - tune.maxrewrite), haproxy will
return HTTP 400 (Bad Request) error. 
Similarly if an HTTP response is larger
than this size, haproxy will return HTTP 502 (Bad Gateway).

32768-8192 = 24576

But we restrict the header value to 4096 . Please correct me if I am wrong. 
RadDeg | x!
-- | --
Inv | sin | ln
π | cos | log
e | tan | √
Ans | EXP | xy

</div></div></div></div></div><div class="gjMOte" style="color: rgb(32, 33, 36); font-family: arial, sans-serif; font-size: medium; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div class="KFFQ0c xKf9F" style="margin-top: 16px; align-items: center; display: flex; margin-bottom: 0px; width: 652px;"><br class="Apple-interchange-newline">

Reply:
To tie up this loose end, the issue is that the user could set multiple headers, or set a header with a dynamic value that expanded to something very large, and thereby exceed the max rewrite buffer size.  Also, the cluster-admin could configure the max rewrite buffer to a smaller size than the default.  Because of dynamic values, the adjustable buffer size, and the multiple involved APIs (Route and IngressController), it is not practical for API validation to ensure that custom headers fit within the buffer.  This issue is now documented under "Failure Modes" in the EP, and the IngressController/Route API godoc for the `spec.httpHeaders.actions`/`value` fields mentions the issue as well.